src/index.js

```diff
@@ -273,8 +280,60 @@ export default function ({ Plugin, types: t }) {
     );
   }
 
+  const statelessVisitorTypes = [
+    'ArrowFunctionExpression',
+    'FunctionDeclaration',
+    'FunctionExpression'
+  ].join('|');
+
   return new Plugin('babel-plugin-react-transform', {
     visitor: {
+      JSXElement(node, parent, scope, file) {
+        if (this.state.isStatelessComponent) {
+          return;
+        }
+
+        if (this.state.isPotentialStatelessComponent) {
+          delete this.state.isPotentialStatelessComponent;
+          this.state.isStatelessComponent = true;
+        }
+      },
+
+      [statelessVisitorTypes]: {
+        enter(node, parent, scope, file) {
```

Comment:
I’m opting to do do some simple internal state tracking on enter to mark as a StatelessComponent and then return the wrapped function on exit.